### PR TITLE
add fit_intercept to LASSO

### LINKED ISSUE #342: LASSO Implementation: Condition bug, missing normalization documentation, and intercept fitting option
## Bug: Incorrect condition for dimension check
https://github.com/smartcorelib/smartcore/blob/0caa8306ff448abbbd1bae7a4fa0dae7af08afcb/src/linear/lasso.rs#L249-L253

The condition `n <= p` is inconsistent with the error message, it should be `n < p`. (my case is full-rank)

## L1 penalty normalization not documented
https://github.com/smartcorelib/smartcore/blob/0caa8306ff448abbbd1bae7a4fa0dae7af08afcb/src/linear/lasso.rs#L12

the penalty in docs is `alpha * sum(abs(beta))`, but I think it's `alpha * n * sum(abs(beta))` in the code, where `n` is the number of observations.

## Feature Request: Option to disable intercept fitting
https://github.com/smartcorelib/smartcore/blob/0caa8306ff448abbbd1bae7a4fa0dae7af08afcb/src/linear/lasso_optimizer.rs#L64

Currently, the mean is always substracted from `y`. Is it possible to add a parameter (e.g., fit_intercept: bool, or more specific substract_mean_from_y: bool) to allow users to disable intercept fitting? (force the `beta_0` parameter in the formula to 0)

### PR DESCRIPTION
**Pull request by georeth**
Fixes #342

### Checklist
- [x] My branch is up-to-date with development branch.
- [x] Everything works and tested on latest stable Rust.
- [x] Coverage and Linting have been applied 

### Current behaviour
LASSO always fits intercept

### New expected behaviour
add `fit_intercept` option

### CODE REVIEW

**Comment by georeth** [comment]
I've checked it converges to the desired value by comparing with sklearn's LassoLars. (the existing test dataset is strong enough that coordinate descent (sklearn's Lasso) converges slow)

I am not very confident about whether the optimizer code assumes `mean(y) == 0` when it computes search direction.

**Comment by Mec-iS** [comment]
if you are not sure, add a couple of tests. You can just translate the ones in sklearn.

**Comment by georeth** [comment]
There are 3 LASSO algorithms involved:
- coordinate descent: fast if data is well-conditioned, but converge slowly otherwise. (sklearn's Lasso)
- LassoLars: compute exact solution, but has numerical issue in rare corner cases, because it needs to compare floating numbers.
- L1LS: Newton interior point method using approximation for speedup (smartcore's implementation)

sklearn's algorithms are not suitable for my use case, and that's why I want to make changes to smartcore.
I tested this PR against sklearn's LassoLars on random-generated data, and the current status of this PR good enough for me.

**Comment by georeth** [comment]
The L1LS paper has no assumption, and it treat LASSO as its special case, so I think it's correct.

**Comment by Mec-iS** [comment]
the docstring at the top of the module `!//` is for noting down these kind of things.
Write there your assumptions with references to the papers like in https://github.com/smartcorelib/smartcore/blob/2bf5f7a1a5db6809ab64c198a8838f773632a148/src/linear/lasso.rs#L4

**Comment by georeth** [comment]
@Mec-iS
The mod doc is really helpful. I had used it to find the L1LS paper:
https://github.com/smartcorelib/smartcore/blob/2bf5f7a1a5db6809ab64c198a8838f773632a148/src/linear/lasso.rs#L20

The doc also referenced a matlab implementation of L1LS (by the paper writer):
https://github.com/smartcorelib/smartcore/blob/2bf5f7a1a5db6809ab64c198a8838f773632a148/src/linear/lasso.rs#L21

The matlab implementation doesn't normalize y at all: https://github.com/cvxgrp/l1_ls/blob/master/l1_ls.m

**Comment by Mec-iS** [comment]
are these changes breaking the API? the existing API will still be accessible or code that exists should be changed to keep using it?

**Comment by georeth** [comment]
It doesn't breaks the API, and it doesn't introduce a numerical diff.
The only diff I can think of is the (de)serialization of LassoParameter, since I introduced a new option.

**Comment by georeth** [comment]
Oh. If the user constructs a LassoParameters directly without using the builder (`default().with_A(...).with_B(...)`), the code breaks.
```
pub struct LassoParameters {
    pub alpha: f64,
    pub normalize: bool,
    pub tol: f64,
    pub max_iter: usize,
    // a new option here
}
```

This also applies to LassoSearchParameters, and LassoSearchParameters doesn't have a builder interface.

**Comment by georeth** [comment]
Should we set intercept to `None` instead of `Some(0.0)` if `fit_intercept` is false?
```rust
pub struct Lasso<TX: FloatNumber + RealNumber, TY: Number, X: Array2<TX>, Y: Array1<TY>> {
    coefficients: Option<X>,
    intercept: Option<TX>,
    _phantom_ty: PhantomData<TY>,
    _phantom_y: PhantomData<Y>,
}
```
I think coefficients and intercept are Option because SupervisedEstimator trait needs a `new()`.

**Comment by Mec-iS** [comment]
0.0 is a value
None is NaN/null
So pick the right one for this case.
If the value is not used/not relevant, then use None.

Breaking changes need to be added to the CHANGELOG.md file

**Review by Mec-iS — APPROVED** [review]

**Comment by Mec-iS** [comment]
Thanks for your contribution
